core/src/main/java/org/apache/calcite/sql2rel/TopDownGeneralDecorrelator.java

```diff
@@ -0,0 +1,1009 @@
+      decorrelateNode = decorrelator.correlateElimination(preparedRel);
+    } catch (UnsupportedOperationException e) {
+      // if the correlation exists in an unsupported operator, retain the original plan.
+    }
+
+    HepProgram postProgram = HepProgram.builder()
+        .addRuleCollection(
+            ImmutableList.of(
+                CoreRules.FILTER_PROJECT_TRANSPOSE,
+                CoreRules.FILTER_INTO_JOIN,
+                CoreRules.MARK_TO_SEMI_OR_ANTI_JOIN_RULE,
+                CoreRules.PROJECT_MERGE,
+                CoreRules.PROJECT_REMOVE))
+        .build();
+    HepPlanner postPlanner = new HepPlanner(postProgram);
+    postPlanner.setRoot(decorrelateNode);
+    return postPlanner.findBestExp();
+  }
+
+  /**
+   * Eliminates Correlate.
+   *
+   * @param rel RelNode
+   * @return  Equivalent RelNode without Correlate
+   */
+  private RelNode correlateElimination(RelNode rel) {
+    if (rel instanceof Correlate) {
+      final Correlate correlate = (Correlate) rel;
+      final RelNode newLeft;
+      if (hasParent) {
+        // if the current decorrelator has a parent, it means that the Correlate must have
+        // correlation from above.
+        assert hasCorrelatedExpressions.containsKey(correlate)
+            && hasCorrelatedExpressions.get(correlate);
+        newLeft = unnest(correlate.getLeft());
+      } else {
+        // otherwise, start a new decorrelation for the left side.
+        newLeft = decorrelateQuery(correlate.getLeft(), builder);
+      }
+
+      // create or update UnnestQuery of left side and corDefs of this decorrelator.
+      UnnestQuery leftInfo = mapRelToUnnestQuery.get(correlate.getLeft());
+      TreeMap<CorDef, Integer> corDefOutputs = new TreeMap<>();
+      Map<Integer, Integer> oldToNewOutputs = new HashMap<>();
+      for (int i = 0; i < correlate.getLeft().getRowType().getFieldCount(); i++) {
+        int newColumnIndex = leftInfo == null ? i : requireNonNull(leftInfo.oldToNewOutputs.get(i));
+        oldToNewOutputs.put(i, newColumnIndex);
+        if (correlate.getRequiredColumns().get(i)) {
+          CorDef corDef = new CorDef(correlate.getCorrelationId(), i);
+          corDefs.add(corDef);
+          corDefOutputs.put(corDef, newColumnIndex);
+        }
+      }
+      if (leftInfo != null) {
+        corDefOutputs.putAll(leftInfo.corDefOutputs);
+      }
+      leftInfo = new UnnestQuery(correlate.getLeft(), newLeft, corDefOutputs, oldToNewOutputs);
+      dedupFreeVarsNode = generateDedupFreeVarsNode(newLeft, leftInfo);
+
+      // decorrelate right side
+      detectCorrelatedExpressions(correlate.getRight());
+      emptyOutputDisable |= correlate.getJoinType() == JoinRelType.MARK;
+      RelNode newRight = unnest(correlate.getRight());
+      UnnestQuery rightInfo = requireNonNull(mapRelToUnnestQuery.get(correlate.getRight()));
+
+      // rewrite condition, adding the natural join condition between the left side and
+      // the domain D that in right side.
```

Comment:
that is produced from the right side.
this the fundamental equation from the paper, shown in Section 3.2